File: 2015/python/day3.py
with open('../input/day3.txt') as file:
    input = file.readline()

coord = [0, 0]
partOneList = []
partOneList.append(coord.copy())

santaCoord = [0, 0]
roboCoord = [0, 0]
partTwoList = []
partTwoList.append(santaCoord.copy())

counter = 0
for direction in input:
    xMod = 0
    yMod = 0

    if direction == '>':
        xMod = 1
    elif direction == '<':
        xMod = -1
    elif direction == '^':
        yMod = 1
    elif direction == 'v':
        yMod = -1
    
    coord[0] += xMod
    coord[1] += yMod

    if counter % 2 == 0:
        santaCoord[0] += xMod
        santaCoord[1] += yMod
    else:
        roboCoord[0] += xMod
        roboCoord[1] += yMod

    if santaCoord not in partTwoList:
        partTwoList.append(santaCoord.copy())

    if roboCoord not in partTwoList:
        partTwoList.append(roboCoord.copy())
    
    if coord not in partOneList:
        partOneList.append(coord.copy())
    
    counter += 1

print('partOne = ' + str(len(partOneList)))
print('partTwo = ' + str(len(partTwoList)))

# $ python3 day3.py
# partOne = 2081
# partTwo = 2341

File: 2015/rust/day3.rs
use std::fs;
use std::collections::HashSet;

#[derive(PartialEq, Hash, Eq, Copy, Clone)]
pub struct Coord {
    x: i32,
    y: i32
}

fn simulate_route(contents: &str, without_robo: bool) -> usize {
    let mut hm: HashSet<Coord> = HashSet::new();
    let mut coord_santa = Coord {x: 0, y: 0};
    let mut coord_robo_santa = Coord {x: 0, y: 0};

    hm.insert(coord_santa);

    let mut coord_ptr;
    for (i, c) in contents.chars().enumerate() {
        if i % 2 == 0 || without_robo {
            coord_ptr = &mut coord_santa;
        } else {
            coord_ptr = &mut coord_robo_santa;
        }

        if c == '^' {
            (*coord_ptr).y += 1;
        } else if c == 'v' {
            (*coord_ptr).y -= 1;
        } else if c == '<' {
            (*coord_ptr).x -= 1;
        } else {
            (*coord_ptr).x += 1;
        }

        hm.insert(*(coord_ptr));
    }
    
    hm.len()
}

fn main() {
    let contents = fs::read_to_string("../input/day3.txt").unwrap();

    println!("partOne = {}", simulate_route(&contents, true));
    println!("partTwo = {}", simulate_route(&contents, false));
}